### PR TITLE
Feature: Masked Dataset

### PR DESCRIPTION
**Pull request by JakobEliasWagner**
# Feature: Masked Dataset

## Description

Not all datasets are consisent in the number of sensors and evaluations. Simulations or measurements are not only performed on a multitude of different grids, but may also contain different numbers of samples in both function spaces/sets. To reflect this the `MaskedOperatorDataset` class is introduced. It is able to handle datasets with this property.

This PR introduces two new operator classes `MaskedOperatorDataset`. The `MaskedOperatorDataset` is able to process datasets with variing number of sensors or evaluations.

### Which issue does this PR tackle?

  - The `OperatorDataset` class is able to only handle uniform evaluation- and sensor-numbers.

### How does it solve the problem?

  - Implements `MaskedOperatorDataset` class to allow for masked sensors and evaluations.

### How are the changes tested?

  - Introduced 10 new unit tests.
  
## Notes

  - The current masking strategy involves to pad all samples to the size of the biggest sample.
    - Other strategies are also possible: cut to the smallest sample by random choice, find a middle-ground between both techniques.
    - In datasets where the numbers vary strongly this can be inefficient.

## Checklist for Contributors

- [x] Scope: This PR tackles exactly one problem.
- [x] Conventions: The branch follows the `feature/title-slug` convention.
- [x] Conventions: The PR title follows the `Bugfix: Title` convention.
- [x] Coding style: The code passes all pre-commit hooks.
- [x] Documentation: All changes are well-documented.
- [x] Tests: New features are tested and all tests pass successfully.
- [x] Changelog: Updated CHANGELOG.md for new features or breaking changes.
- [x] Review: A suitable reviewer has been assigned.

## Checklist for Reviewers:

- [ ] The PR solves the issue it claims to solve and only this one.
- [ ] Changes are tested sufficiently and all tests pass.
- [ ] Documentation is complete and well-written.
- [ ] Changelog has been updated, if necessary.

### CODE REVIEW

File: src/continuiti/data/dataset.py [hunk 1]
```diff
@@ -131,29 +126,214 @@ def __getitem__(
...
-        sample = {"x": x, "u": u, "y": y, "v": v}
+        assert type(member) is type(
+            values
+        ), f"All types of tensors in one space need to match. But found {type(member)} and {type(values)}."
+
+        ndim: int
+        if mask is not None:
+            assert isinstance(
+                member, torch.Tensor
+            ), f"When providing a mask the member and values need to be tensors. But found {type(member)}"
```
Comment (samuelburbulla) [inline src/continuiti/data/dataset.py:251]:
```suggestion
            assert isinstance(member, torch.Tensor), (
                f"When providing a mask the member and values need to be tensors, but found {type(member)}")
```

File: src/continuiti/data/dataset.py [hunk 2]
```diff
@@ -131,29 +126,214 @@ def __getitem__(
...
         Returns:
-            Input samples with class transformations applied.
+            A boolean value True when the space description is valid.
         """
-        sample = {"x": x, "u": u, "y": y, "v": v}
+        assert type(member) is type(
+            values
+        ), f"All types of tensors in one space need to match. But found {type(member)} and {type(values)}."
+
+        ndim: int
```
Comment (samuelburbulla) [inline src/continuiti/data/dataset.py:247]:
```suggestion
```

File: src/continuiti/data/dataset.py [hunk 3]
```diff
@@ -131,29 +126,214 @@ def __getitem__(
...
+                [di.size(0) == member[0].size(0) for di in member[1:]]
+            ), "Dimensions of all samples of the member need to match."
+            assert all(
+                [vi.size(0) == values[0].size(0) for vi in values[1:]]
+            ), "Dimensions of all function values need to match."
+            ndim = member[0].dim()
+
+        assert (
+            ndim == 2
+        ), f"{self.__class__.__name__} currently only supports exactly one dim and one size dimension."
```
Comment (samuelburbulla) [inline src/continuiti/data/dataset.py:264]: Why is that? So I cannot have shape of (batch_size, 3, 128, 128)?

File: src/continuiti/data/dataset.py [hunk 4]
```diff
@@ -131,29 +126,214 @@ def __getitem__(
...
+
+        return True
+
+    def _pad_list_space(
+        self, member: List[torch.Tensor], values: List[torch.Tensor]
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Transforms a space described by lists of tensors (image) to a space described by padded tensors and a mask.
```
Comment (samuelburbulla) [inline src/continuiti/data/dataset.py:271]: (image)?

File: src/continuiti/data/dataset.py [hunk 5]
```diff
@@ -131,29 +126,214 @@ def __getitem__(
...
+        Args:
+            member: List of tensors describing the locations of the samples.
+            values: List of tensors describing the function values of the samples.
+
+        Returns:
+            padded member tensor, padded values tensor, and matching mask.
+        """
+        assert not any(
+            [torch.any(torch.isinf(mi)) for mi in member]
+        ), "Expects domain to be truncated in finite space."
```
Comment (samuelburbulla) [inline src/continuiti/data/dataset.py:282]: Why is this assertion necessary? Someone might come up with a good reason for using infs in the data, do we have to prevent that?

Comment (samuelburbulla) [reply src/continuiti/data/dataset.py:282]: Ah, I see we're using inf for padding. However, does it hurt to have more infs (non-masked) in the dataset?

File: src/continuiti/data/dataset.py [hunk 6]
```diff
@@ -131,29 +126,214 @@ def __getitem__(
...
+
+        member_padded = pad_sequence(
+            [mi.transpose(0, 1) for mi in member],
+            batch_first=True,
+            padding_value=torch.inf,
+        ).transpose(1, 2)
+        values_padded = pad_sequence(
+            [vi.transpose(0, 1) for vi in values], batch_first=True, padding_value=0
```
Comment (samuelburbulla) [inline src/continuiti/data/dataset.py:290]:
```suggestion
            [vi.transpose(0, 1) for vi in values],
            batch_first=True,
            padding_value=0,
```

Comment (samuelburbulla) [reply src/continuiti/data/dataset.py:290]: Why do we pad once with inf and once with 0? Seems arbitrary

File: src/continuiti/data/dataset.py [hunk 7]
```diff
@@ -131,29 +126,214 @@ def __getitem__(
...
+        ).transpose(1, 2)
+
+        mask = member_padded != torch.inf
+        member_padded[
+            ~mask
+        ] = 0  # mask often applied by adding a tensor with -inf values in masked locations (e.g. in scaled dot product).
```
Comment (samuelburbulla) [inline src/continuiti/data/dataset.py:296]: What is different here if why just used 0 for padding in l. 287?

File: src/continuiti/data/dataset.py [hunk 8]
```diff
@@ -131,29 +126,214 @@ def __getitem__(
...
-        for dim, val in sample.items():
-            if dim in self.transform:
-                sample[dim] = self.transform[dim](val)
+        if self.opt_mask is not None:
+            opt_mask = self.opt_mask[idx]
+        else:
+            opt_mask = None
 
-        return sample["x"], sample["u"], sample["y"], sample["v"]
+        return tensors[0], tensors[1], tensors[2], tensors[3], ipt_mask, opt_mask
```
Comment (samuelburbulla) [inline src/continuiti/data/dataset.py:339]:
```suggestion
        return *tuple(tensors), ipt_mask, opt_mask
```

File: tests/data/test_dataset.py [hunk 9]
```diff
@@ -32,3 +37,150 @@ def test_dataset():
...
+            ),
+            v_transform=Normalize(
+                torch.ones(self.v_dim, 1) / 2.0,
+                torch.ones(self.v_dim, 1) * torch.sqrt(torch.tensor([1 / 12])),
+            ),
+        )
+        dataloader = DataLoader(dataset, batch_size=self.batch_size)
+
+        for x, u, y, v, ipt_mask, opt_mask in dataloader:
+            assert True
```
Comment (samuelburbulla) [inline tests/data/test_dataset.py:186]: Can we do more here?
